File: src/schema/events.ts
/**
 * The SSE envelope and the closed set of event names.
 *
 * The invariant: **events are display transport, never the source of truth.**
 * `seq` is a process-monotonic counter that resets on server restart, so it is a
 * dedupe key and NOT a resume cursor (plan §6.16). A reconnecting client re-fetches
 * `GET /sessions/:id` and reconciles by message id; nothing replays from a seq.
 * The practical rule that falls out: any state a client cannot rebuild from a
 * fresh fetch is a bug in the event design, not something to fix with replay.
 *
 * The name list is closed here so the TUI store can switch on it exhaustively.
 * Payload shapes are declared as types rather than validated on the wire: the bus
 * hands listeners the object it published, in-process, so a Zod parse on the SSE
 * boundary would only re-check what the emitter already typed. The envelope IS
 * parsed, because the client reads it off a socket.
 */
import { z } from "zod";
import type {
  AskQuestion,
  BackgroundJob,
  Message,
  Part,
  Session,
  Turn,
  TurnStatus,
  WorkflowAgent,
  WorkflowRun,
} from "./parts.ts";

// ---- the closed event-name set ---------------------------------------------

/**
 * Spec §3's event list, plus `tool.log`.
 *
 * `tool.log` is not named in spec §3 but is required by spec §5.3 — "console.*
 * lines stream live to the UI **and** batch into the tool result". Streaming
 * console output has no other carrier: `message.delta` is model text, and the
 * batched copy only reaches the client when the tool result lands. It is declared
 * here rather than left for M3 to add, because this list is frozen.
 */
export const EVENT_TYPES = [
  "session.created",
  "session.updated",
  "session.activity",
  "message.started",
  "message.delta",
  "message.part",
  "message.finished",
  "message.retry",
  "tool.log",
  "turn.finished",
  "ask.question",
  "job.spawned",
  "job.exited",
  "workflow.updated",
  "workflow.agent",
  "workflow.log",
] as const;

export const EventType = z.enum(EVENT_TYPES);
export type EventType = z.infer<typeof EventType>;

// ---- the envelope ----------------------------------------------------------

/**
 * Every event carries a process-monotonic `seq` and a `ts`, both stamped by the
 * bus at publish time. `sessionId` is what `GET /events?sessionId=` filters on;
 * events with no session (none today) reach every subscriber.
 *
 * `data` is `unknown` on the envelope on purpose — its shape is per-`type`, and
 * the typed view is `BoughEventOf<T>` below. Validating `data` generically here
 * would mean one schema that is wrong for fifteen payloads.
 */
export const BoughEvent = z.object({
  type: EventType,
  sessionId: z.string().optional(),
  seq: z.number(),
  ts: z.number(),
  data: z.unknown(),
});
export type BoughEvent = z.infer<typeof BoughEvent>;

// ---- per-event payloads ----------------------------------------------------

/** `message.delta` — incremental model text for a streaming message. */
export interface MessageDeltaData {
  messageId: string;
  delta: string;
}

/** `message.part` — one finalized Part appended to a message. */
export interface MessagePartData {
  messageId: string;
  part: Part;
}

/** `message.finished` — the message is complete; `pending` is now false. */
export interface MessageFinishedData {
  messageId: string;
}

/**
 * `message.retry` — the round is being re-attempted (a tool call whose input was
 * cut off mid-stream, or a transient provider failure). The message re-streams
 * from the top, so a client drops its streaming buffer for it. `reason` says which
 * so the UI can show what happened rather than a bare spinner (spec §5 Retry).
 */
export interface MessageRetryData {
  messageId: string;
  /** 1-based; retries are bounded and an exhausted one surfaces as a turn error. */
  attempt: number;
  reason: string;
}

/**
 * `tool.log` — one `console.*` line from a running program, keyed to the
 * `tool_call` that produced it. Display-only: the same line is also batched into
 * the tool result the model receives, so dropping these changes nothing about
 * what the model sees.
 */
export interface ToolLogData {
  messageId: string;
  callId: string;
  line: string;
}

/**
 * `session.activity` — a cheap-tier blurb describing what the session is doing
 * right now. Fails silently and never blocks a turn; `activity: null` clears it
 * (spec §12, plan §6.11).
 */
export interface SessionActivityData {
  sessionId: string;
  activity: string | null;
}

/** `turn.finished` — emitted after `message.finished`, once per turn. */
export interface TurnFinishedData {
  turnId: string;
  sessionId: string;
  status: TurnStatus;
  /** Present when status is `error` — names the limit or the failure (spec §5). */
  error?: string;
}

/** `job.spawned` / `job.exited` — background shells, per spec §9. */
export type JobSpawnedData = BackgroundJob;
export type JobExitedData = BackgroundJob;

/** `workflow.log` — one narrator `log()` line from a running script. */
export interface WorkflowLogData {
  runId: string;
  line: string;
}

/**
 * The payload each event name carries. This is the map the TUI store reduces
 * over; a new arm here is a compile error in every exhaustive switch.
 */
export interface EventDataMap {
  "session.created": Session;
  "session.updated": Session;
  "session.activity": SessionActivityData;
  "message.started": Message;
  "message.delta": MessageDeltaData;
  "message.part": MessagePartData;
  "message.finished": MessageFinishedData;
  "message.retry": MessageRetryData;
  "tool.log": ToolLogData;
  "turn.finished": TurnFinishedData;
  "ask.question": AskQuestion;
  "job.spawned": JobSpawnedData;
  "job.exited": JobExitedData;
  "workflow.updated": WorkflowRun;
  "workflow.agent": WorkflowAgent;
  "workflow.log": WorkflowLogData;
}

/** A stamped event narrowed to one name — `BoughEventOf<"message.delta">`. */
export type BoughEventOf<T extends EventType> = Omit<BoughEvent, "type" | "data"> & {
  type: T;
  data: EventDataMap[T];
};

/** The discriminated union of every stamped event. */
export type AnyBoughEvent = { [K in EventType]: BoughEventOf<K> }[EventType];

/**
 * An event as published — everything except the bus-assigned stamp. The bus
 * (T0.3) assigns `seq`/`ts` and returns the stamped event.
 */
export type EventInput = { [K in EventType]: Omit<BoughEventOf<K>, "seq" | "ts"> }[EventType];

/** Re-exported so consumers of the turn's outcome need one import. */
export type { Turn };

File: crates/bough-core/src/schema/events.rs
//! The SSE envelope and the closed set of event names (port of
//! `src/schema/events.ts`).
//!
//! The invariant: **events are display transport, never the source of truth.**
//! `seq` is a process-monotonic counter that resets on server restart, so it is
//! a dedupe key and NOT a resume cursor. A reconnecting client re-fetches
//! `GET /sessions/:id` and reconciles by message id; nothing replays from a seq.
//!
//! The name list is closed here so the TUI store can match on it exhaustively
//! (no default arm — a new event type must be a compile error). The envelope IS
//! parsed (it comes off a socket); payloads are typed but not wire-validated.

use serde::{Deserialize, Serialize};
use serde_json::Value;

use super::parts::{Part, TurnStatus};

// ---- the closed event-name set ---------------------------------------------

/// The 16 wire names, frozen. `tool.log` is deliberately included: streaming
/// console output has no other carrier.
pub const EVENT_TYPES: [&str; 16] = [
    "session.created",
    "session.updated",
    "session.activity",
    "message.started",
    "message.delta",
    "message.part",
    "message.finished",
    "message.retry",
    "tool.log",
    "turn.finished",
    "ask.question",
    "job.spawned",
    "job.exited",
    "workflow.updated",
    "workflow.agent",
    "workflow.log",
];

/// The closed 16-name enum. Reducers match exhaustively with no default arm.
#[derive(Serialize, Deserialize, Clone, Copy, Debug, PartialEq, Eq, Hash)]
pub enum EventType {
    #[serde(rename = "session.created")]
    SessionCreated,
    #[serde(rename = "session.updated")]
    SessionUpdated,
    #[serde(rename = "session.activity")]
    SessionActivity,
    #[serde(rename = "message.started")]
    MessageStarted,
    #[serde(rename = "message.delta")]
    MessageDelta,
    #[serde(rename = "message.part")]
    MessagePart,
    #[serde(rename = "message.finished")]
    MessageFinished,
    #[serde(rename = "message.retry")]
    MessageRetry,
    #[serde(rename = "tool.log")]
    ToolLog,
    #[serde(rename = "turn.finished")]
    TurnFinished,
    #[serde(rename = "ask.question")]
    AskQuestion,
    #[serde(rename = "job.spawned")]
    JobSpawned,
    #[serde(rename = "job.exited")]
    JobExited,
    #[serde(rename = "workflow.updated")]
    WorkflowUpdated,
    #[serde(rename = "workflow.agent")]
    WorkflowAgent,
    #[serde(rename = "workflow.log")]
    WorkflowLog,
    /// A hook did something. Announced so the TUI can say so as it happens —
    /// a hook that silently rewrites a command is indistinguishable from a
    /// harness that is behaving strangely.
    #[serde(rename = "hook.fired")]
    HookFired,
}

impl EventType {
    pub fn as_str(&self) -> &'static str {
        match self {
            EventType::SessionCreated => "session.created",
            EventType::SessionUpdated => "session.updated",
            EventType::SessionActivity => "session.activity",
            EventType::MessageStarted => "message.started",
            EventType::MessageDelta => "message.delta",
            EventType::MessagePart => "message.part",
            EventType::MessageFinished => "message.finished",
            EventType::MessageRetry => "message.retry",
            EventType::ToolLog => "tool.log",
            EventType::TurnFinished => "turn.finished",
            EventType::AskQuestion => "ask.question",
            EventType::JobSpawned => "job.spawned",
            EventType::JobExited => "job.exited",
            EventType::WorkflowUpdated => "workflow.updated",
            EventType::WorkflowAgent => "workflow.agent",
            EventType::WorkflowLog => "workflow.log",
            EventType::HookFired => "hook.fired",
        }
    }
}

// ---- the envelope ----------------------------------------------------------

/// Every event carries a process-monotonic `seq` and a `ts`, both stamped by
/// the bus at publish time. `sessionId` is what `GET /events?sessionId=`
/// filters on; events with no session reach every subscriber.
///
/// `data` is untyped on the envelope on purpose — its shape is per-`type`;
/// payloads are deserialized per-type by the consumer.
#[derive(Serialize, Deserialize, Clone, Debug, PartialEq)]
pub struct BoughEvent {
    #[serde(rename = "type")]
    pub r#type: EventType,
    #[serde(rename = "sessionId", default, skip_serializing_if = "Option::is_none")]
    pub session_id: Option<String>,
    pub seq: u64,
    pub ts: i64,
    #[serde(default)]
    pub data: Value,
}

/// An event as published — everything except the bus-assigned stamp. The bus
/// assigns `seq`/`ts` and returns the stamped [`BoughEvent`].
#[derive(Clone, Debug, PartialEq)]
pub struct EventInput {
    pub r#type: EventType,
    pub session_id: Option<String>,
    pub data: Value,
}

// ---- per-event payloads ----------------------------------------------------
// `session.created`/`session.updated` carry a `Session`; `message.started` a
// `Message`; `ask.question` an `AskQuestion`; `job.spawned`/`job.exited` a
// `BackgroundJob`; `workflow.updated` a `WorkflowRun`; `workflow.agent` a
// `WorkflowAgent`. The rest are below.

/// `message.delta` — incremental model text for a streaming message.
#[derive(Serialize, Deserialize, Clone, Debug, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct MessageDeltaData {
    pub message_id: String,
    pub delta: String,
}

/// `message.part` — one finalized Part appended to a message.
#[derive(Serialize, Deserialize, Clone, Debug, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct MessagePartData {
    pub message_id: String,
    pub part: Part,
}

/// `message.finished` — the message is complete; `pending` is now false.
#[derive(Serialize, Deserialize, Clone, Debug, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct MessageFinishedData {
    pub message_id: String,
}

/// `message.retry` — the round is being re-attempted. The message re-streams
/// from the top, so a client drops its streaming buffer for it.
#[derive(Serialize, Deserialize, Clone, Debug, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct MessageRetryData {
    pub message_id: String,
    /// 1-based; retries are bounded and an exhausted one surfaces as a turn error.
    pub attempt: u32,
    pub reason: String,
}

/// `tool.log` — one `console.*` line from a running program, keyed to the
/// `tool_call` that produced it. Display-only.
#[derive(Serialize, Deserialize, Clone, Debug, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct ToolLogData {
    pub message_id: String,
    pub call_id: String,
    pub line: String,
}

/// `session.activity` — what the session is doing right now, in two
/// independent slots: the cheap-tier `activity` blurb ("running the test
/// suite") and the `command` a shell verb is blocked on right now
/// (`cargo test`). Both render on the busy line.
///
/// TWO PUBLISHERS, ONE EVENT, so each field is `Option<Option<_>>`: **absent
/// leaves the other slot alone, `null` clears it.** The cheap tier and
/// `hostfn::shell` publish concurrently and neither knows the other's state
/// — a plain `Option` would make every blurb erase the running command and
/// every command erase the blurb. Absent-is-untouched is what keeps them
/// independent without a shared owner.
#[derive(Serialize, Deserialize, Clone, Debug, Default, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct SessionActivityData {
    pub session_id: String,
    #[serde(
        default,
        deserialize_with = "double_option",
        skip_serializing_if = "Option::is_none"
    )]
    pub activity: Option<Option<String>>,
    #[serde(
        default,
        deserialize_with = "double_option",
        skip_serializing_if = "Option::is_none"
    )]
    pub command: Option<Option<String>>,
}

/// Absent → `None`, `null` → `Some(None)`, value → `Some(Some(v))`.
///
/// REQUIRED, not decoration: serde's derive collapses an explicit `null` to
/// `None` for a plain `Option<Option<T>>`, which is exactly the case that
/// carries the CLEAR. Without this the busy line named a finished command
/// forever — the fields would have been write-only.
pub fn double_option<'de, D, T>(deserializer: D) -> Result<Option<Option<T>>, D::Error>
where
    D: serde::Deserializer<'de>,
    T: Deserialize<'de>,
{
    Deserialize::deserialize(deserializer).map(Some)
}

/// `turn.finished` — emitted after `message.finished`, once per turn.
#[derive(Serialize, Deserialize, Clone, Debug, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct TurnFinishedData {
    pub turn_id: String,
    pub session_id: String,
    pub status: TurnStatus,
    /// Present when status is `error` — names the limit or the failure.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

/// `workflow.log` — one narrator `log()` line from a running script.
#[derive(Serialize, Deserialize, Clone, Debug, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct WorkflowLogData {
    pub run_id: String,
    pub line: String,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn event_type_list_has_exactly_16_names_and_round_trips() {
        assert_eq!(EVENT_TYPES.len(), 16);
        for name in EVENT_TYPES {
            let t: EventType = serde_json::from_value(Value::String(name.into())).unwrap();
            assert_eq!(t.as_str(), name);
            assert_eq!(serde_json::to_value(t).unwrap(), Value::String(name.into()));
        }
    }

    #[test]
    fn unknown_event_type_rejected_on_the_envelope() {
        let json = r#"{"type":"message.exploded","seq":1,"ts":2,"data":{}}"#;
        assert!(serde_json::from_str::<BoughEvent>(json).is_err());
    }

    #[test]
    fn envelope_round_trips() {
        let json = r#"{"type":"message.delta","sessionId":"s1","seq":7,"ts":123,"data":{"messageId":"m1","delta":"hi"}}"#;
        let e: BoughEvent = serde_json::from_str(json).unwrap();
        assert_eq!(e.r#type, EventType::MessageDelta);
        assert_eq!(e.session_id.as_deref(), Some("s1"));
        let d: MessageDeltaData = serde_json::from_value(e.data.clone()).unwrap();
        assert_eq!(d.delta, "hi");
        let out = serde_json::to_string(&e).unwrap();
        assert!(out.contains(r#""type":"message.delta""#));
        assert!(out.contains(r#""sessionId":"s1""#));
    }
}
